### DISCUSSION TITLE
Parse error with json containing several key/value structures

### DISCUSSION THREAD

**Discussion by edu-arr99** [nlohmann/json#2919] [Q&A] [answered]
Hi! I'm doing a university assignment regarding graphs and I'm trying to parse a json file containing several key/value structures using a class called Parser. Just to be clear, I have to do this using said class:
```
#include <iostream>
#include <fstream>
using namespace std;
using json = nlohmann::json;

class Parser
{
private:
    string path;                // file path
    json jsonGraph;                     
public:
    Parser(string path_);

    void clear(); // Clears parser saved atributes

    void readJSON(); // Parses JSON file and saves data into class


    void uGraphMake(UnDirectedGraph<string, double> &tempGraph); // Adds the parsed data into the specified undirected graph

    void dGraphMake(DirectedGraph<string, double> &tempGraph); // Adds the parsed data into the specified directed graph
};
```

At first it's pretty much straight forward, with the constructor and the readJSON() method: 
```
Parser::Parser(string path_) {
    this->path = path_;
}

void Parser::readJSON() {
    ifstream file(this->path, ifstream::in);
    this->jsonGraph = json::parse(file);
}
```

I've already verified that I'm opening the file correctly. This json file contains several key/value structures, like this: 
```
{
    {
        "City": "Juliaca",
        "DBTZ": "U",
        "Name": "Juliaca",
        "Country": "Peru",
        "IATA/FAA": "JUL",
        "Longitude": "-70.158169",
        "ICAO": "SPJL",
        "Airport ID": "2792",
        "Latitude": "-15.467103",
        "Timezone": "12552",
        "DST": "-5",
        "destinations": [
            "2802",
            "2812",
            "2789"
        ]
    },
    {
        "City": "Tumbes",
        "DBTZ": "U",
        "Name": "Pedro Canga",
        "Country": "Peru",
        "IATA/FAA": "TBP",
        "Longitude": "-80.381356",
        "ICAO": "SPME",
        "Airport ID": "2796",
        "Latitude": "-3.552528",
        "Timezone": "115",
        "DST": "-5",
        "destinations": [
            "2789"
        ]
    }
}
```
I'm getting this error:
```
[json.exception.parse_error.101] parse error at line 2, column 5: syntax error while parsing object key - unexpected '{'; expected string literal
```
I understand why I'm getting this error message, but I don't know how to solve it. I have to keep each structure separated, since each airport is going to be a node, but it won't let me put braces directly after the opening braces, and if I take out the opening and closing braces of the whole file, it stops reading after the first "}". What could I do? Is there even a way I could parse this file as a nlohmann::json object? Or does anyone have any ideas? :(

**Comment by edu-arr99** [comment, marked as answer]
Just in case anyone needs it, I found a solution:
```
void Parser::readJSON() {
    ifstream file(this->path, ifstream::in);        // open file
    string line, jsonString;
    while (getline(file, line))                     // we pass the json info into a string first
        jsonString += line + "\n";
    
    this->jsonGraph = json::parse(jsonString);            // we parse from the string
}
```
I replaced the global curly braces for square brackets in the json file, and managed to make it work this way.

**Reply by gregmarr** [reply]
> I replaced the global curly braces for square brackets in the json file, and managed to make it work this way.

That was your main problem, you had a malformed JSON file.  After that, it should read fine.

    void Parser::readJSON() {
        ifstream file(this->path);
        file >> this->jsonGraph;
    }

**Reply by edu-arr99** [reply]
Oh no it doesn't, I got a similar error:
```
[json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing object key - unexpected end of input; expected '{',  '[' or string literal

```
It had to be assigned to a string for it to work
